File: nodo-x402/sdk/python/nodo_x402/solana.py
"""
Solana Wallet Integration for NODO x402
"""
import json
from typing import Optional, Union
from pathlib import Path

from .exceptions import WalletError, InsufficientFunds


class SolanaWallet:
    """
    Solana wallet for USDC payments.
    
    Supports:
    - Loading keypair from file
    - Loading from private key bytes
    - USDC SPL token transfers
    """
    
    USDC_MINT = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"
    USDC_DECIMALS = 6
    
    def __init__(
        self,
        keypair_path: Optional[str] = None,
        private_key: Optional[bytes] = None,
        rpc_url: str = "https://api.mainnet-beta.solana.com",
    ):
        self.rpc_url = rpc_url
        self._keypair = None
        self._client = None
        self._address = None
        
        # Load keypair
        if keypair_path:
            self._load_keypair_from_file(keypair_path)
        elif private_key:
            self._load_keypair_from_bytes(private_key)
    
    @property
    def address(self) -> Optional[str]:
        """Get wallet public address."""
        return self._address
    
    def _load_keypair_from_file(self, path: str):
        """Load keypair from JSON file."""
        try:
            from solders.keypair import Keypair
            
            expanded_path = Path(path).expanduser()
            with open(expanded_path) as f:
                secret = json.load(f)
            
            self._keypair = Keypair.from_bytes(bytes(secret))
            self._address = str(self._keypair.pubkey())
            
        except ImportError:
            # Solana SDK not installed, use mock
            self._address = "MockWallet111111111111111111111111111111111"
        except Exception as e:
            raise WalletError(f"Failed to load keypair: {e}")
    
    def _load_keypair_from_bytes(self, private_key: bytes):
        """Load keypair from bytes."""
        try:
            from solders.keypair import Keypair
            
            self._keypair = Keypair.from_bytes(private_key)
            self._address = str(self._keypair.pubkey())
            
        except ImportError:
            self._address = "MockWallet111111111111111111111111111111111"
        except Exception as e:
            raise WalletError(f"Failed to load keypair: {e}")
    
    async def get_usdc_balance(self) -> float:
        """Get USDC balance."""
        if not self._keypair:
            return 0.0
        
        try:
            from solana.rpc.async_api import AsyncClient
            from solders.pubkey import Pubkey
            
            if not self._client:
                self._client = AsyncClient(self.rpc_url)
            
            pubkey = self._keypair.pubkey()
            response = await self._client.get_token_accounts_by_owner(
                pubkey,
                {"mint": Pubkey.from_string(self.USDC_MINT)}
            )
            
            if response.value:
                total = 0.0
                for account in response.value:
                    amount = account.account.data.parsed["info"]["tokenAmount"]["uiAmount"]
                    total += float(amount)
                return total
            
            return 0.0
            
        except ImportError:
            return 100.0  # Mock balance for development
        except Exception as e:
            raise WalletError(f"Failed to get balance: {e}")
    
    async def send_usdc(
        self,
        recipient: str,
        amount: float,
        memo: str = "",
    ) -> str:
        """
        Send USDC to recipient.
        
        Args:
            recipient: Recipient wallet address
            amount: Amount in USDC (decimal)
            memo: Optional memo
            
        Returns:
            Transaction signature
        """
        if not self._keypair:
            raise WalletError("No keypair loaded")
        
        try:
            from solana.rpc.async_api import AsyncClient
            from solana.transaction import Transaction
            from solders.pubkey import Pubkey
            from solders.system_program import transfer, TransferParams
            from spl.token.instructions import (
                transfer_checked,
                TransferCheckedParams,
                get_associated_token_address,
            )
            from spl.memo.instructions import create_memo, MemoParams
            
            if not self._client:
                self._client = AsyncClient(self.rpc_url)
            
            # Convert amount to lamports (USDC has 6 decimals)
            amount_lamports = int(amount * (10 ** self.USDC_DECIMALS))
            
            # Get token accounts
            sender_pubkey = self._keypair.pubkey()
            recipient_pubkey = Pubkey.from_string(recipient)
            usdc_mint = Pubkey.from_string(self.USDC_MINT)
            
            sender_token_account = get_associated_token_address(
                sender_pubkey, usdc_mint
            )
            recipient_token_account = get_associated_token_address(
                recipient_pubkey, usdc_mint
            )
            
            # Build transaction
            tx = Transaction()
            
            # Add memo if provided
            if memo:
                tx.add(create_memo(
                    MemoParams(
                        signer=sender_pubkey,
                        message=memo.encode()
                    )
                ))
            
            # Add transfer instruction
            tx.add(transfer_checked(
                TransferCheckedParams(
                    program_id=Pubkey.from_string("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"),
                    source=sender_token_account,
                    mint=usdc_mint,
                    dest=recipient_token_account,
                    owner=sender_pubkey,
                    amount=amount_lamports,
                    decimals=self.USDC_DECIMALS,
                )
            ))
            
            # Send transaction
            response = await self._client.send_transaction(
                tx,
                self._keypair,
            )
            
            if response.value:
                return str(response.value)
            else:
                raise WalletError("Transaction failed")
            
        except ImportError:
            # Mock transaction for development
            import hashlib
            import time
            mock_sig = hashlib.sha256(
                f"{recipient}{amount}{memo}{time.time()}".encode()
            ).hexdigest()
            return mock_sig
            
        except Exception as e:
            raise WalletError(f"Transaction failed: {e}")
    
    async def close(self):
        """Close wallet connections."""
        if self._client:
            await self._client.close()

File: nodo-x402/sdk/typescript/src/solana.ts
/**
 * Solana Wallet Integration for NODO x402
 */

import { WalletError } from './errors';

interface SolanaWalletOptions {
  keypair?: Uint8Array;
  rpcUrl?: string;
}

const USDC_MINT = 'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v';
const USDC_DECIMALS = 6;

export class SolanaWallet {
  private keypair: unknown = null;
  private connection: unknown = null;
  private rpcUrl: string;
  private _address: string | null = null;

  constructor(options: SolanaWalletOptions = {}) {
    this.rpcUrl = options.rpcUrl || 'https://api.mainnet-beta.solana.com';

    if (options.keypair) {
      this.loadKeypair(options.keypair);
    }
  }

  get address(): string | null {
    return this._address;
  }

  private loadKeypair(secretKey: Uint8Array): void {
    try {
      // Dynamic import to avoid issues if @solana/web3.js not installed
      const { Keypair } = require('@solana/web3.js');
      this.keypair = Keypair.fromSecretKey(secretKey);
      this._address = (this.keypair as { publicKey: { toBase58(): string } }).publicKey.toBase58();
    } catch {
      // Fallback for development without Solana SDK
      this._address = 'MockWallet111111111111111111111111111111111';
    }
  }

  async getUsdcBalance(): Promise<number> {
    if (!this.keypair) {
      return 0;
    }

    try {
      const { Connection, PublicKey } = require('@solana/web3.js');
      const { getAssociatedTokenAddress, getAccount } = require('@solana/spl-token');

      if (!this.connection) {
        this.connection = new Connection(this.rpcUrl);
      }

      const ownerPubkey = (this.keypair as { publicKey: unknown }).publicKey;
      const usdcMint = new PublicKey(USDC_MINT);

      const tokenAccount = await getAssociatedTokenAddress(usdcMint, ownerPubkey);

      try {
        const account = await getAccount(this.connection, tokenAccount);
        return Number(account.amount) / Math.pow(10, USDC_DECIMALS);
      } catch {
        return 0; // Token account doesn't exist
      }
    } catch {
      // Return mock balance for development
      return 100;
    }
  }

  async sendUsdc(recipient: string, amount: number, memo = ''): Promise<string> {
    if (!this.keypair) {
      throw new WalletError('No keypair loaded');
    }

    try {
      const { Connection, PublicKey, Transaction } = require('@solana/web3.js');
      const {
        getAssociatedTokenAddress,
        createTransferCheckedInstruction,
        TOKEN_PROGRAM_ID,
      } = require('@solana/spl-token');

      if (!this.connection) {
        this.connection = new Connection(this.rpcUrl);
      }

      const kp = this.keypair as {
        publicKey: unknown;
        secretKey: Uint8Array;
      };

      const usdcMint = new PublicKey(USDC_MINT);
      const recipientPubkey = new PublicKey(recipient);

      // Get token accounts
      const senderTokenAccount = await getAssociatedTokenAddress(
        usdcMint,
        kp.publicKey
      );
      const recipientTokenAccount = await getAssociatedTokenAddress(
        usdcMint,
        recipientPubkey
      );

      // Build transaction
      const tx = new Transaction();

      // Add memo if provided
      if (memo) {
        const { createMemoInstruction } = require('@solana/spl-memo');
        tx.add(createMemoInstruction(memo, [kp.publicKey]));
      }

      // Add transfer instruction
      const amountInLamports = BigInt(Math.round(amount * Math.pow(10, USDC_DECIMALS)));

      tx.add(
        createTransferCheckedInstruction(
          senderTokenAccount,
          usdcMint,
          recipientTokenAccount,
          kp.publicKey,
          amountInLamports,
          USDC_DECIMALS,
          [],
          TOKEN_PROGRAM_ID
        )
      );

      // Send transaction
      const conn = this.connection as {
        sendTransaction(tx: unknown, signers: unknown[]): Promise<string>;
      };

      const { Keypair } = require('@solana/web3.js');
      const signer = Keypair.fromSecretKey(kp.secretKey);

      const signature = await conn.sendTransaction(tx, [signer]);
      return signature;
    } catch (error) {
      // Mock transaction for development
      if (String(error).includes('Cannot find module')) {
        const crypto = require('crypto');
        return crypto.randomBytes(32).toString('hex');
      }
      throw new WalletError(`Transaction failed: ${error}`);
    }
  }
}


